### PR TITLE
[Fix] Fix Inaccurate `num_training_steps` in Lr Scheduler

### LINKED ISSUE #6137: Inaccurate `num_train_steps` in the LR scheduler during multi-GPU parallelism.
### Describe the bug

When I was building my own LoRa training code, I encountered an issue with the `num_training_steps` parameter in the LR scheduler. I attempted to follow the official training script from [diffusers](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L787-L827). But I observed that the variable `args.max_train_steps` is not taking into account the division by the number of GPUs after the `accelerator.prepare` in the instantiation of the LR scheduler. This results in the `lr_scheduler.num_training_steps` being greater than the actual number of training steps. Consequently, this prevents adjustment strategies like cosine annealing from reaching 0. More precisely, the instantiation of the LR scheduler should derive `args.max_train_steps` from [here](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L825) then multiply the `accelerator.num_processes`. 

d1ce6ae876d63c57b69f022d58c9a13

I use `get_cosine_with_hard_restarts_schedule_with_warmup` with `num_cycles=10`.

I found that the author of the consistency model seems to have identified this issue. Therefore, they handle the length of the dataloader for multiple GPUs before the `accelerator.prepare`. You can see the relevant code here: https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_sd_wds.py#L1025.

The attribute `train_dataloader.num_batches` takes into account the multi-GPU scenario during its calculation.

### Reproduction

    overrode_max_train_steps = False
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / gradient_accumulation_steps)
    if max_train_steps == -1:
        max_train_steps = num_train_epochs * num_update_steps_per_epoch
        overrode_max_train_steps = True

    lr_scheduler = get_cosine_with_hard_restarts_schedule_with_warmup(
        optimizer, num_warmup_steps=lr_warmup_steps * accelerator.num_processes,
        num_training_steps=max_train_steps * accelerator.num_processes,
        num_cycles=lr_num_cycles
    )

    lora_net, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
        lora_net, optimizer, train_dataloader, lr_scheduler
    )

    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / gradient_accumulation_steps)
    if overrode_max_train_steps:
        max_train_steps = num_train_epochs * num_update_steps_per_epoch
    num_train_epochs = math.ceil(max_train_steps / num_update_steps_per_epoch)

### Logs

_No response_

### System Info

None

### Who can help?

 @sayakpaul @patrickvonplaten @yiyixuxu @patrickvonplaten @yiyixuxu @DN6 @sayakpaul @patrickvonplaten

### PR DESCRIPTION
**Pull request by Dai-Wenxun**
# What does this PR do?

Fixes #6137

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul

### CODE REVIEW

**Review by sayakpaul — APPROVED** [review]
Thank you!

**Comment by sayakpaul** [comment]
Could you please resolve the code quality issues?

**Comment by sayakpaul** [comment]
Could we please fix the code quality issues so that this PR can be merged?

**Comment by sayakpaul** [comment]
A gentle bump.

**Comment by Dai-Wenxun** [comment]
Ok

**Comment by sayakpaul** [comment]
Hey there, 

Could you please fix the code quality issues?

**Comment by sayakpaul** [comment]
Hey @Dai-Wenxun a gentle ping here.

**Comment by sayakpaul** [comment]
@Dai-Wenxun we're still interested in the PR? Could you maybe resolve the conflicts and fix the code formatting issues?
